batch_finder_resource_method.md

```diff
@@ -0,0 +1,418 @@
+
+BATCH FINDER should not have any visible side effects.
+For example, it should be safe to call whenever the client wishes.
+However, this is not something enforced by the framework, and it is up to the application developer that there are no side effects.
+
+It is important to note that:
+- the operations may execute on the server out of order
+- the response objects are expected to be returned in the same order and position as the respective input search criteria.
+- The BATCH_FINDER will require implementing a resource method to handle a BATCH_FINDER requests. It won't behave like a multiplexer that will call automatically existing finders
+
+## Protocol
+### Request
+#### URI
+The URI templates below assume variables with types as follows:
+
+    batch_finder : simple string identifying a batch_finder method name
+    resource : simple string identifying a resource 
+    search_criteria : simple string identifying the criteria filter name
+
+| Resource                           | URI Template          | Example                   | Method | Semantics                    |
+| ---------------------------------- | --------------------- | ------------------------- | ------ | ---------------------------- |
+| Collection, Association            | {resource}?bq={batch_finder}&{search_criteria}=| /PhotoResource?bq=searchPhotos&photoCriteria=List((id:1, format:JPG),(id:2, format:BMP)) | GET    | invokes the specified batch_finder |
+
+At least, 2 query parameters will have to be set for a batch finder:
+
+- The "bq" query parameter is reserved for passing the batch finder method name
+- A second query parameter will be used to pass a set of different search criteria. The name of this query parameter is set in the [BatchFinder method annotation](/rest.li/batch_finder_resource_method#method-annotation-and-parameters).
+For example, with @BatchFinder(value="findUsers", batchParam="batchCriteria"), the batch query parameter name is "batchCriteria". 
+The type of this query parameter is a List.
+
+Eg.
+
+    curl "d2://userSearchResults?bq=findUsers&batchCriteria=List((firstName:pauline, age:12),(lastName:iglou))" -X GET
+
+The other query parameters will be applied as common filters across all batch requests.
+
+Here is an example batch request with two individual finders using the following criteria:
+
+- filter by first name and age
+- filter by last name and age
+
+Eg.
+
+    curl "d2://userSearchResults?bq=findUsers&batchCriteria=List((firstName:pauline),(lastName:iglou))&age=21" -X GET
+
+#### Pagination support
+##### 1) Common pagination for all search criteria  
+The developer can pass additional parameters to specify a common pagination. It will be more efficient than adding a pagination context inside each criteria object.  
+Eg.
+
+    curl "d2://userSearchResults?q=findUsers&batchCriteria=List((firstName:pauline, age:12),(lastName:iglou))&firstName=max&start=10&count=10" -X GET
```

Comment:
Can we also have the examples with the protocol V2?

Reply:
This is the example with Protocol V2 and I will add a V1.

Reply:
The "curl" request won't work with d2. For protocol V2, you need to set the Header.
For example
`curl --header "X-RestLi-Protocol-Version: 2.0.0"  "http://localhost:10546/userSearchResults?q=findUsers&batchCriteria=List((firstName:pauline, age:12),(lastName:iglou))&firstName=max&start=10&count=10"`